lib/Runtime/Types/PathTypeHandler.cpp

```diff
@@ -299,11 +299,90 @@ namespace Js
+        newInlineSlots[i + 1] = auxSlotOne;
+
+        Assert(i == this->GetInlineSlotCapacity());
+    }
+
+    BOOL PathTypeHandlerBase::DeleteLastProperty(DynamicObject *const object)
+    {
+        // We are deleting the last property from PathTypeHandler
+        DynamicType* predecessorType = this->GetPredecessorType();
+
+        // -----------------------------------------------------------------------------------------
+        //     Current Type  | Predecessor Type |                  Action
+        // -----------------------------------------------------------------------------------------
+        //          OHI      |      OHI         | No movement needed
+        // -----------------------------------------------------------------------------------------
+        //          OHI      |    Not OHI       | Not possible (Should be a BUG)
+        // -----------------------------------------------------------------------------------------
+        //        Not OHI    |    Not OHI       | No movement needed
+        // -----------------------------------------------------------------------------------------
+        //        Not OHI    |      OHI         | Move from auxSlots to inline slots (ReOptimize)
+        // -----------------------------------------------------------------------------------------
+
+        bool isCurrentTypeOHI = this->IsObjectHeaderInlinedTypeHandlerUnchecked();
+
+        Assert(predecessorType->GetTypeHandler()->IsPathTypeHandler());
+        PathTypeHandlerBase* predecessorTypeHandler = (PathTypeHandlerBase*)predecessorType->GetTypeHandler();
+
+        Assert(predecessorTypeHandler->GetUnusedBytesValue() == (this->GetUnusedBytesValue() - 1));
+
+        bool isPredecessorTypeOHI = predecessorTypeHandler->IsObjectHeaderInlinedTypeHandlerUnchecked();
+
+        AssertMsg(!isCurrentTypeOHI || isPredecessorTypeOHI, "Current type is ObjectHeaderInlined but previous type is not ObjectHeaderInlined?");
+
+        if (!isCurrentTypeOHI && isPredecessorTypeOHI)
+        {
+            Assert(predecessorTypeHandler->GetInlineSlotCapacity() == (this->GetUnusedBytesValue() - 1));
+            this->MoveAuxSlotsToObjectHeader(object);
+        }
+
+#if DBG
+        if ((isCurrentTypeOHI && isPredecessorTypeOHI) || (!isCurrentTypeOHI && !isPredecessorTypeOHI))
+        {
+            AssertMsg(this->GetInlineSlotCapacity() == predecessorTypeHandler->GetInlineSlotCapacity(), "InlineSlotCapacity of types should match");
+        }
 #endif
+
+        Assert(predecessorTypeHandler->GetSlotCapacity() <= this->GetSlotCapacity());
+
+        // Another type (this) reached the old (predecessorType) type so share it.
+        // ShareType will take care of invalidating fixed fields and removing singleton object from predecessorType
+        predecessorType->ShareType();
+
+        this->typePath->ClearSingletonInstanceIfSame(object);
+
+        object->ReplaceTypeToPredecessorType(predecessorType);
+
+        return TRUE;
```

Comment:
1 scenario was if the property doesn't exist, but will be good to check other scenarios from spec.

---
In reply to: [86213097](https://github.com/Microsoft/ChakraCore/pull/1897#discussion_r86213097) [](ancestors = 86213097,86039231)